### PR TITLE
feat(tba-integration): [nan-1259] support netsuite tba authorization

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
Adds support for the TBA flow employed by netsuite

## Issue ticket number and link
NAN-1259

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/frontend/lib/index.ts [hunk 1]
```diff
@@ -128,7 +130,11 @@ export default class Nango {
             return this.customAuth(providerConfigKey, connectionId, this.convertCredentialsToConfig(credentials), connectionConfig);
         }
 
-        const url = this.hostBaseUrl + `/oauth/connect/${providerConfigKey}${this.toQueryString(connectionId, options as ConnectionConfig)}`;
+        let url = this.hostBaseUrl + `/oauth/connect/${providerConfigKey}${this.toQueryString(connectionId, options as ConnectionConfig)}`;
+
+        if (options && 'credentials' in options && 'token_id' in options.credentials && 'token_secret' in options.credentials) {
+            url = this.hostBaseUrl + `/auth/tba/${providerConfigKey}${this.toQueryString(connectionId, options as ConnectionConfig)}`;
+        }
```
Comment (TBonnin) [inline packages/frontend/lib/index.ts:137]: I feel the logic and condition above should leave in credentials, for example a`isCustomAuth()` and `urlPath()` functions would significantly simplify the code in `nango.auth()`

File: packages/server/lib/controllers/auth/tba.ts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        connectionId,
+        params: {
+            oauth_token: oauth_token as string,
+            state: session.id.replace(/-/g, '')
+        },
+        connectionConfig
+    });
+
+    res.redirect(redirectUrl.toString());
+});
```
Comment (TBonnin) [inline packages/server/lib/controllers/auth/tba.ts:235]: a lot of the code above can fail and throw. what kind of response message is sent back to the user in this case?

File: packages/server/lib/helpers/tba.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import type { OAuthSession } from '@nangohq/shared';
+import querystring from 'querystring';
+import * as crypto from 'node:crypto';
+import type { Template as ProviderTemplate, IntegrationConfig as ProviderConfig } from '@nangohq/types';
+import { interpolateStringFromObject } from '@nangohq/shared';
+
+export function percentEncode(str: string) {
+    return encodeURIComponent(str)
+        .replace(/[!'()*]/g, (c) => '%' + c.charCodeAt(0).toString(16).toUpperCase())
+        .replace(/%20/g, '%20');
```
Comment (TBonnin) [inline packages/server/lib/helpers/tba.ts]: are we replacing anything here?

File: packages/server/lib/helpers/tba.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    encodedParams.sort((a, b) => {
+        if (a[0] === b[0]) {
+            return a[1] < b[1] ? -1 : 1;
+        }
+        return a[0] < b[0] ? -1 : 1;
+    });
+
+    return encodedParams.map((pair) => pair.join('=')).join('&');
```
Comment (TBonnin) [inline packages/server/lib/helpers/tba.ts]: you can chain everything I think 
```
export function encodeParameters(allParams: Record<string, string>): string {
    return Object.entries(allParams)
        .map(([key, value]) => [percentEncode(key), percentEncode(value)] as [string, string])
        .sort(([a, aVal], [b, bVal]) => {
            if (a === b) {
                return aVal < bVal ? -1 : 1;
            }
            return a < b ? -1 : 1;
        })
        .map((pair) => pair.join('='))
        .join('&');
}
```

File: packages/frontend/lib/index.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/frontend/lib/index.ts:137]: Yeah, agreed. Let me get this out and refactor this some tomorrow.

File: packages/server/lib/controllers/auth/tba.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/server/lib/controllers/auth/tba.ts:235]: The generic error which wouldn't be helpful. I'll add in more error reporting and catching but for now this is going out for one customer with one use case so will make it more robust tomorrow.

**Comment by khaliqgant** [comment]
As mentioned in the comments this is an initial implementation with a need to refactor, but it is a bit difficult to test locally w/o verifying with the client first. So will ship this now and revisit tomorrow.

**Pull request merged**